src/observer/CameraSensorServer.cpp

```diff
@@ -0,0 +1,439 @@
+      }
+    }
+    if(client_was_alive)
+    {
+      client_was_alive = false;
+      mc_rtc::log::critical("[CameraSensorServer::{}] Client disconnected", name_);
+    }
+    // Wait for the client to become available
+    data_->server_idle.notify_all();
+    std::this_thread::sleep_for(std::chrono::milliseconds(50));
+  }
+}
+
+void CameraSensorServer::do_computation()
+{
+  if(data_->skip)
+  {
+    return;
+  }
+  {
+    const std::lock_guard<std::mutex> lock(points_mtx_);
+    ground_points_ = points_;
+    corrected_ground_points_.reserve(ground_points_.size());
+  }
+  {
+    mc_rtc::log::info("[Step 1] Estimate the z - pitch to bring back to 0");
+    auto start = std::chrono::high_resolution_clock::now();
+    double pitch = 0.;
+    double t_z = 0.;
+
+    ceres::Problem problem;
+    for(const auto & point : ground_points_)
+    {
+      const sva::PTransformd X_0_p(point);
+      const sva::PTransformd X_s_p = X_0_p * (data_->X_b_s * data_->X_0_b).inv();
+      ceres::CostFunction * cost_function = PitchZCostFunctor::Create(X_s_p, data_->X_0_b, data_->X_b_s);
+      problem.AddResidualBlock(cost_function, new ceres::CauchyLoss(0.5), &pitch, &t_z);
+    }
+    ceres::CostFunction * min_p = new ceres::AutoDiffCostFunction<Minimize, 1, 1>(new Minimize());
+    problem.AddResidualBlock(min_p, nullptr, &pitch);
+    ceres::CostFunction * min_z = new ceres::AutoDiffCostFunction<Minimize, 1, 1>(new Minimize());
+    problem.AddResidualBlock(min_z, nullptr, &t_z);
+
+    ceres::Solver::Options options;
+    options.linear_solver_type = ceres::DENSE_QR;
+    ceres::Solver::Summary summary;
+    ceres::Solve(options, &problem, &summary);
+
+    const sva::PTransformd X_b_b(sva::RotY(pitch).cast<double>(), Eigen::Vector3d(0., 0., t_z));
+
+    // Compute the 3D point in world frame
+    {
+      corrected_ground_points_.clear();
+      for(const auto & pp : ground_points_)
+      {
+        // From camera frame to world frame
+        sva::PTransformd X_0_p(pp);
+        sva::PTransformd X_s_p = X_0_p * (data_->X_b_s * data_->X_0_b).inv();
+        X_0_p = X_s_p * data_->X_b_s * X_b_b * data_->X_0_b;
+        corrected_ground_points_.push_back(X_0_p.translation());
+      }
+    }
+    auto stop = std::chrono::high_resolution_clock::now();
+    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(stop - start);
+    mc_rtc::log::warning("[Step 1] It tooks {} microseconds -> {} milliseconds", duration.count(),
+                         duration.count() / 1000.);
+  }
+  {
+    mc_rtc::log::info("[Step 2] Perform ICP");
+    auto start = std::chrono::high_resolution_clock::now();
+
+    pc_estimated_ground_points_->points_ = corrected_ground_points_;
+
+    if(pc_full_ground_reconstructed_points_->points_.empty())
+    {
+      *pc_full_ground_reconstructed_points_ = *pc_estimated_ground_points_;
+      ipc::scoped_lock<ipc::interprocess_mutex> lck(data_->points_mtx);
+      data_->points.resize(0);
+      return;
+    }
```

Comment:
Seconde note: I'm a little unsure about this here but I guess the intent is to measure the displacement of the ground so this is a way to ensure we have initial data to compute the displacement?